src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Async.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Internal.IL;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    public partial class CompilerTypeSystemContext
+    {
+        public MethodDesc GetAsyncVariantImplMethod(MethodDesc asyncMetadataMethod)
+        {
+            MethodDesc asyncMetadataMethodDef = asyncMetadataMethod.GetTypicalMethodDefinition();
+            MethodDesc result = _asyncVariantImplHashtable.GetOrCreateValue((EcmaMethod)asyncMetadataMethodDef);
+
+            if (asyncMetadataMethodDef != asyncMetadataMethod)
+            {
+                TypeDesc owningType = asyncMetadataMethod.OwningType;
+                if (owningType.HasInstantiation)
+                    result = GetMethodForInstantiatedType(result, (InstantiatedType)owningType);
+
+                if (asyncMetadataMethod.HasInstantiation)
+                    result = GetInstantiatedMethod(result, asyncMetadataMethod.Instantiation);
+            }
+
+            return result;
+        }
+
+        private sealed class AsyncVariantImplHashtable : LockFreeReaderHashtable<EcmaMethod, AsyncVariantImplMethod>
```

Comment:
Are we going with the plan to make these more durable and not just transient per compilation (like unboxing stubs)?

Just curious what tipped us into this direction (I am fine with it).